NETProvider/src/FirebirdSql.Data.FirebirdClient/Client/Common/XsqldaMarshaler.cs

```diff
@@ -153,75 +132,104 @@ public IntPtr MarshalManagedToNative(Charset charset, Descriptor descriptor)
 			return MarshalManagedToNative(xsqlda, xsqlvar);
 		}
 
-		public IntPtr MarshalManagedToNative(XSQLDA xsqlda, XSQLVAR[] xsqlvar)
+		public static IntPtr MarshalManagedToNative(XSQLDA xsqlda, XSQLVAR[] xsqlvar)
 		{
-			int size = ComputeLength(xsqlda.sqln);
-			IntPtr ptr = Marshal.AllocHGlobal(size);
+			var size = ComputeLength(xsqlda.sqln);
+			var ptr = Marshal.AllocHGlobal(size);
 
 			Marshal.StructureToPtr(xsqlda, ptr, true);
 
-			for (int i = 0; i < xsqlvar.Length; i++)
+			for (var i = 0; i < xsqlvar.Length; i++)
 			{
-				int offset = ComputeLength(i);
+				var offset = ComputeLength(i);
 				Marshal.StructureToPtr(xsqlvar[i], GetIntPtr(ptr, offset), true);
 			}
 
 			return ptr;
 		}
 
-		public Descriptor MarshalNativeToManaged(Charset charset, IntPtr pNativeData)
+		public static Descriptor MarshalNativeToManaged(Charset charset, IntPtr pNativeData)
 		{
 			return MarshalNativeToManaged(charset, pNativeData, false);
 		}
 
-		public Descriptor MarshalNativeToManaged(Charset charset, IntPtr pNativeData, bool fetching)
+		static int sqlTypeOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqltype").AsInt();
+		static int sqlscaleOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqlscale").AsInt();
+		static int sqlsubtypeOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqlsubtype").AsInt();
+		static int sqllenOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqllen").AsInt();
+		static int sqldataOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqldata").AsInt();
+		static int sqlindOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqlind").AsInt();
+		static int sqlname_lengthOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqlname_length").AsInt();
+		static int sqlnameOffset = Marshal.OffsetOf(typeof(XSQLVAR), "sqlname").AsInt();
+		static int relname_lengthOffset = Marshal.OffsetOf(typeof(XSQLVAR), "relname_length").AsInt();
+		static int relnameOffset = Marshal.OffsetOf(typeof(XSQLVAR), "relname").AsInt();
+		static int ownername_lengthOffset = Marshal.OffsetOf(typeof(XSQLVAR), "ownername_length").AsInt();
+		static int ownernameOffset = Marshal.OffsetOf(typeof(XSQLVAR), "ownername").AsInt();
+		static int aliasname_lengthOffset = Marshal.OffsetOf(typeof(XSQLVAR), "aliasname_length").AsInt();
+		static int aliasnameOffset = Marshal.OffsetOf(typeof(XSQLVAR), "aliasname").AsInt();
+
+
+		static void MarshalVAR(IntPtr ptr, XSQLVAR var)
 		{
-			// Obtain XSQLDA information
-			XSQLDA xsqlda = new XSQLDA();
+			unsafe
```

Comment:
I still don't get what are you going to do with `UnmanagedMemoryStream`. :) Maybe because I think the below code is OK. It's readable. And it's just one place.